Workspace displays status bar

DTasev opened the pull request:
**Description of work.**
Adds a status bar to the workspace display. This will display messages regarding the status of user actions - e.g. copy completed, operation in progress.

All of that functionality has been refactored into common classes:
- `status_bar_view.py` provides the parent window that contains the status bar.
- `user_notifier.py` provides the functions for notifying the user via tooltip **(Windows only)**, and status bar
- `data_copier.py` provides functions for copying the data from the matrix/table workspaces

- Note that the code in `data_copier.py` and  `user_notifier.py` is copied from the old `table_copying.py`(which has now been deleted), and turned into two classes with unit tests 🎉 .
- Workspace displays are now in `mantidqt.widgets.workspacedisplay` package

**To test:**
- Copy something (bin, spectra or selected cells), the status bar should show "Copy Successful"
- Open a workspace with a lot of spectra and bins, select the _bins_, and copy them. The status bar will show "Operation in progress" if you have selected enough bins (as copying bins is slower than spectra)

<!-- Instructions for testing. -->

Fixes #24630 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

martyngigg commented:
@DTasev It looks like we've got some conflicts here now.

martyngigg commented on qt/python/mantidqt/widgets/commonworkspacedisplay/status_bar_view.py:
```diff
@@ -0,0 +1,24 @@
+        self.status_bar = QStatusBar(self)
+        self.setStatusBar(self.status_bar)
+
+        self.resize(600, 400)
```
Just out of curiosity why is resize necessary?

DTasev replied:
Qt seems to have some idea of the minimum size, but that appears pretty small on 1920x1080. Giving a size of 600x400 makes them pretty much the same size as Mantid Plot and I like the consistency between Mantid Plot and the Workbench

martyngigg replied:
:+1: for the consistency. Should we make the default size a constructor parameter so that it can be easily overridden so it wouldn't have to be resized twice?

DTasev commented:
I think grouping them together is an awesome idea! I will make the change as part of this PR

DTasev commented:
That should be fixed now - `StatusBarView`'s closedEvent is now called when the window is closed. The event closes the central widget (which deletes the ADS observer), then closes itself.

The change required a rename in the InstrumentView to fit the new call. I checked and the `InstrumentView` and plots are still deleted properly.

martyngigg commented on qt/applications/workbench/workbench/plotting/test/test_figuremanager.py:
```diff
@@ -1,4 +1,16 @@
+
+from qtpy.QtWidgets import QApplication
+
+from mantidqt.widgets.workspacedisplay.test_helper.qt_widget_finder import QtWidgetFinder
```
Would this be better in the `mantidqt.utils.qt.test` package as I think it's more general?

DTasev replied:
Good idea

DTasev commented:
I've fixed some renaming issues from merging with master, and removed the FigureManager deletion test to a new issue #24727 - because it passes if ran via PyCharm, fails if ran via CTest, and the widgets are properly deleted in the workbench (confirmed via running the test manually)

DTasev commented:
Removed the `instrumentview_view` test from being tested against qt4, hopefully that should make everything pass

DTasev commented:
All the tests have passed now, could you have another look @martyngigg?

martyngigg commented on qt/applications/workbench/workbench/plotting/figurewindow.py:
```diff
@@ -19,7 +19,7 @@
 # local imports
 from mantidqt.plotting.figuretype import FigureType, figure_type
-from mantidqt.widgets.common.observing_view import ObservingView
+from mantidqt.widgets.workspacedisplay.observing_view import ObservingView
```
Really minor but perhaps `ObservingView` should be outside of `workspacedisplay` now that we are using it elsewhere? That way the packages are not coupled. It could just go to `mantidqt.widgets` ?

DTasev replied:
`ads_observer` and `observing_presenter` are also used outside the `workspacedisplay`, maybe moving them into a `common` module would be better? Or an `observing` module?

martyngigg replied:
Yes I think that would be good. Maybe call it `observers` with those 3 classes in? If it expands over time it could become a package later.

martyngigg commented on qt/python/mantidqt/widgets/workspacedisplay/status_bar_view.py:
```diff
@@ -0,0 +1,45 @@
+    close_signal = Signal()
+    rename_signal = Signal(str)
+
+    def __init__(self, parent, central_widget, name, window_width=600, window_height=400, presenter=None):
```
`window_size` as a `tuple` rather than separate parameters?

DTasev replied:
I prefer separate parameters, I'm not sure we gain much from using a tuple, because then it's not immediately clear which is used as height and width

martyngigg commented on qt/widgets/mplcpp/src/Zoomer.cpp:
```diff
@@ -21,10 +21,9 @@ const char *TOOLBAR_MODE_ZOOM = "zoom rect";
   bool showCoordinates(false);
-  return Python::Object(
-      backend.attr(TOOLBAR_CLS)(canvas->pyobj(), parent, showCoordinates));
+  return Python::Object(backend.attr(TOOLBAR_CLS)(
```
Nice catch on the memory leak!
